#' Read Bismark M-Bias file
#' 
#' This function inputs a M-Bias file generated by bismark_methylation_extractor
#' and returns a data frame (long format) with all contexts and strands found.
#' 
#' @param file Path to Bismark M-Bias file
#'
#' @return data frame with the following columns:
#' * Context: CpG, CHG or CHH
#' * Strand: Forward or Reverse
#' * Cycle: Sequencing cycle
#' * Count.Methylated:
#' * Count.Unmethylated:
#' * Percent.Methylation:
#' * Coverage: 
#' 
#' @export
#' @md
read_bismark_mbias <- function(file) {
  data <- readr::read_lines(file)
  
  end_table <- which(data == "")
  start_table <- c(1, head(end_table, -1) + 1)
  
  purrr::map2_df(start_table, end_table, function(start, end) {
    match <- stringr::str_match(data[start], "(\\w+) context( \\(R(\\d)\\))?")
    context <- match[1, 2]
    strand <- ifelse(match[1, 4] == "1", "Forward", "Reverse")
    
    readr::read_tsv(data[start:end], skip = 2) %>%
      dplyr::rename(Cycle = position, Count.Methylated = "count methylated", 
             Count.Unmethylated = "count unmethylated",
             Percent.Methylation = "% methylation", Coverage = coverage) %>%
      tibble::add_column(Strand = strand, Context = context, .before = "Cycle")
  })
}

#' Plot M-Bias
#' 
#' This function inputs a data frame generated by read_bismark_mbias function
#' and plots M-Bias using ggplot2.
#' 
#' If there are Strand information (Forward and Reverse), this function plots a
#' Strand as facet.
#'
#' @param mbias data frame must contains the folowing columns:
#' * Context
#' * Strand (filled with NA if single-end data)
#' * Cycle
#' * Percent.Methylation
#'
#' @return ggplot2 object
#' @export
#' @md
plot_mbias <- function(mbias) {
  p <- mbias %>%
    ggplot(aes(x = Cycle, y = Percent.Methylation, color = Context)) +
      geom_line()
  
  if (! all(is.na(mbias$Strand))) 
    p <- p + facet_wrap(~ Strand)
  
  p
}

#' Read Genomic Nucleotide Frequencies
#' 
#' The 'genomic_nucleotide_frequencies.txt' can be generated using either
#' `bismark_genome_preparation --genomic_composition` or `bam2nuc` Bismark
#' commands.
#'
#' @param file path to genomic_nucleotide_frequencies.txt
#'
#' @return data frame with the following columns:
#' * Bases: (di-)nucleotides
#' * Count: frequency in genome
#' 
#' @export
#' @md
#'
read_bismark_genomic_nucleotide_frequencies <- function(file) {
  readr::read_tsv(file, col_names = c("Bases", "Count")) %>%
    dplyr::mutate(Type = ifelse(Bases %in% c("A", "C", "G", "T"), "Nucleotide", "Dinucleotide")) %>%
    dplyr::group_by(Type) %>%
    mutate(Frequency = Count / sum(Count)) %>%
    arrange(desc(Type))
}

#' Plot genomic nucleotide frequencies
#'
#' @param genomic_nucleotide_frequencies data frame should have the columns:
#' * Bases
#' * Type
#' * Count
#'
#' @return
#' @export
#'
plot_genomic_nucleotide_frequencies <- function(genomic_nucleotide_frequencies, show = c("Nucleotide", "Dinucleotide")) {
  genomic_nucleotide_frequencies %>%
    filter(Type %in% show) %>%
    ggplot(aes(x = reorder(Bases, desc(Type)), y = Count)) + 
      geom_col() +
      labs(x = "Bases")
}

#' Read Bismark nucleotide coverage
#'
#' @param file 
#'
#' @return
#' @export
read_bismark_nucleotide_coverage <- function(file) {
  readr::read_tsv(file) %>%
    dplyr::rename("Bases" = "(di-)nucleotide", 
      "Count.Sample" = "count sample",
      "Percent.Sample" = "percent sample", 
      "Count.Genomic" = "count genomic", 
      "Percent.Genomic" = "percent genomic", 
      "Coverage" = "coverage")  %>%
    dplyr::mutate(Type = ifelse(Bases %in% c("A", "C", "G", "T"), "Nucleotide", "Dinucleotide"))
}


#' Plot Bismark nucleotide coverage
#'
#' @param nucleotide_coverage 
#' @param show 
#'
#' @return
#' @export
plot_bismark_nucleotide_coverage <- function(nucleotide_coverage, show = c("Nucleotide", "Dinucleotide")) {
  nucleotide_coverage %>%
    filter(Type %in% show) %>%
    select(Bases, Type, Sample = Percent.Sample, Genomic = Percent.Genomic) %>%
    gather(From, Percent, -Bases, -Type) %>%
    ggplot(aes(x = reorder(Bases, desc(Type)), y = Percent, fill = From)) +
      geom_col(position = "dodge") + 
    labs(x = "Bases", y = "%")
}

#' Read Bismark alignment report file
#'
#' @param file Path to Bismark alignment report file
#'
#' @return data frame with the following columns:
#' * Reads:
#' * Unique.Hits:
#' * No.Hits:
#' * Multiple.Hits:
#' * Not.Extracted:
#' * OT:
#' * OB:
#' * CTOT:
#' * CTOB:
#' * C:
#' * CpG.Methylated:
#' * CHG.Methylated:
#' * CHH.Methylated:
#' * Unknown.Methylated:
#' * CpG.Unmethylated:
#' * CHG.Unmethylated:
#' * CHH.Unmethylated:
#' * Unknown_Unmethylated:
#' 
#' @export
#' @md
read_bismark_align_report <- function(file) {
  tibble(text = read_file(file)) %>%
    mutate(
      Reads = str_match(., "total:\\t(\\d+)\\n")[2],
      Unique.Hits = str_match(., "alignments:\\t(\\d+)\\n")[2],
      No.Hits = str_match(., "condition:\\t(\\d+)\\n")[2],
      Multiple.Hits = str_match(., "uniquely:\\t(\\d+)\\n")[2],
      Not.Extracted = str_match(., "extracted:\\t(\\d+)\\n")[2],
      OT = str_match(., "CT/CT:\\t(\\d+)\\t")[2],
      OB = str_match(., "CT/GA:\\t(\\d+)\\t")[2],
      CTOT = str_match(., "GA/CT:\\t(\\d+)\\t")[2],
      CTOB = str_match(., "GA/GA:\\t(\\d+)\\t")[2],
      C = str_match(., "analysed:\\t(\\d+)\\n")[2],
      CpG.Methylated = str_match(., "methylated.+CpG.+:\\t(\\d+)\\n")[2],
      CHG.Methylated = str_match(., "methylated.+CHG.+:\\t(\\d+)\\n")[2],
      CHH.Methylated = str_match(., "methylated.+CHH.+:\\t(\\d+)\\n")[2],
      Unknown.Methylated = str_match(., "methylated.+Unknown.+:\\t(\\d+)\\n")[2],
      CpG.Unmethylated = str_match(., "unmethylated.+CpG.+:\\t(\\d+)\\n")[2],
      CHG.Unmethylated = str_match(., "unmethylated.+CHG.+:\\t(\\d+)\\n")[2],
      CHH.Unmethylated = str_match(., "unmethylated.+CHH.+:\\t(\\d+)\\n")[2],
      Unknown_Unmethylated = str_match(., "unmethylated.+Unknown.+:\\t(\\d+)\\n")[2]
    ) %>%
    select(-text) %>%
    mutate(across(everything(), as.numeric))
}
